### ISSUE TITLE
(Matlab) Wrong syntax highlighting for numeric literals

### ISSUE THREAD

**Issue by stefano-zanotti-88** [ScintillaOrg/lexilla#120]
When using Notepad++ v8.4.7 (I don't know which version of lexilla it uses), and editing a file with Matlab syntax highlighting, the numeric literals are not properly colored:
```
d    = 123;
x    = 0x123ABC;
b    = 0b010101;
xs64 = 0x2As64;
xs32 = 0x2As32;
xs16 = 0x2As16;
xs8  = 0x2As8;
xu64 = 0x2Au64;
xu32 = 0x2Au32;
xu16 = 0x2Au16;
xu8  = 0x2Au8;
bs64 = 0b10s64;
bs32 = 0b10s32;
bs16 = 0b10s16;
bs8  = 0b10s8;
bu64 = 0b10u64;
bu32 = 0b10u32;
bu16 = 0b10u16;
bu8  = 0b10u8;
c = .1;
c = 1.1;
c = .1e1;
c = 1.1e1;
c = 1e1;
c = 1i;
c = 1j;
c = .1e2j;
c = 1e2j;
```
![img](https://user-images.githubusercontent.com/7887495/206847556-21828151-64c1-49a9-8e8c-5bab2c3c363b.png)

For non-decimal literals, only the leading 0 is colored, whereas the whole literal should be - maybe including the optional suffix.
The suffixes above (eight for fixed-width integers, and two for imaginary numbers) are all the ones I know about.
There are no suffixes for 128-bit literals, nor a prefix for octal.

For the documentation, see here:
https://www.mathworks.com/help/matlab/matlab_prog/specify-hexadecimal-and-binary-numbers.html
I've tried all the literals shown above, and indeed all of them are recognized by MATLAB, though Matlab's editor doesn't highlight numbers.

Octave seems to support the prefixes as well:
https://docs.octave.org/v6.3.0/Numeric-Data-Types.html
but I don't know about its support for the suffixes.

Regarding complex literals: any decimal or floating point literal - but not binary/hex literals - can have an "i" or "j" suffix, which makes the number imaginary.
Complex numbers are just the sum of a real part and an imaginary part; they're an expression rather than a literal, so they don't need to be treated specially.
Also, a plain "i" or "j" evaluates to "1i", but since they can be reassigned as regular variables, I wouldn't color them as numeric literals. E.g.
```
a = i; % a == 1i
b = 1i; % b == 1i
i = 1;
a = i; % a == 1  "surprising"
b = 1i; % b == 1i  still working
c = 0x23i % causes "Error: Invalid digit in hexadecimal literal. Supported hexadecimal digits are 0-9 and A-F. Supported type suffixes are u8, u16, u32, u64, and s8, s16, s32, s64."
```
Despite the error message, hex literals also accept lowercase a-f.
The suffix and prefix are both case-insensitive too.

Then there's the topic of "constants" (which are miscellaneous "literals" - though more properly they are re-assignable variables that come pre-bound to default values):
```
eps
Inf
inf
intmax
intmin
namelengthmax
realmax
realmin
pi
NaN
nan
NaT
false
true
```
with the added complication that some but not all of them are also functions that can be used to build matrices, e.g.
```
n = nan(2, 3, 'single');
f = false(5, 2);
p = pi(3,4); % error "Too many input arguments."
```
and also the fact that they are case-sensitive:
```
k = inf;
k = +inf;
k = -inf;
k = Inf;
k = +Inf;
k = -Inf;
k = inF; % error
k = INF; % error
k = InF; % error
t = NaT; % ok, it's "not-a-datetime"
t = nat; % error
```
All of these are reassignable too:
```
a = inf; % a is infinity
inf = 5;
a = inf; % a == 5
a = Inf; % a is infinity
```
So it's probably better to just treat them as regular identifiers.

In my opinion, the correct behavior for lexilla should be to either 1) color the whole literal without the suffix, or 2) color the whole literal including the suffix, but not a plain "i" or "j".

You might want to look here for a more detailed description of Matlab's grammar:
https://github.com/mathworks/MATLAB-Language-grammar/blob/master/Matlab.tmbundle/Syntaxes/MATLAB.tmLanguage

**Comment by rdipardo** [comment]
@stefano-zanotti-88,
> In my opinion, the correct behavior for lexilla should be to either 1) color the whole literal without the suffix . . .

[Notepad2's implementation][0] goes a bit further in satisfying 1), at least for hexadecimal literals:

![MATLAB-literals-np2-trunk](https://user-images.githubusercontent.com/59004801/206878677-cf75d0a8-b8c4-4411-8af7-6b1aec02d5c9.png)

> . . . or 2) color the whole literal including the suffix, but not a plain "i" or "j"

Something similar was [recently implemented for PowerShell][1], so it's easily achievable. 

I don't know MATLAB, but in general I would say that annotated literals are still atomic values. With the exception of postfix operators, a suffix, by definition, is not an independent language element.

[0]: https://github.com/zufuliu/notepad2/blob/main/scintilla/lexers/LexMatlab.cxx
[1]: https://github.com/ScintillaOrg/lexilla/blob/621fefc5333752ce0fb2ea0d3df62c1d09b4c836/lexers/LexPowerShell.cxx#L53

**Comment by zufuliu** [comment]
A simple "fix" is allow arbitrary letter (Octave supports underscore as separator, and uppercase I and J as suffix) in `SCE_MATLAB_NUMBER` (similar to C++ lexer).

**Comment by stefano-zanotti-88** [comment]
Yes, I think that it could be a good solution to generically match the following as numeric literals (the first one for floating points -- which I guess is already working correctly -- the second one for integers with arbitrary prefixes/suffixes):
```
(\.[0-9]+|[0-9]+\.[0-9]*)([eE][-+]?[0-9]+)?|[0-9]+[eE][-+]?[0-9]+
[0-9][0-9_a-zA-Z]*
```
After all, rejecting wrong suffixes is more of a semantic check than a syntactic one.
Are floating point literals a different token from integer literals in lexilla? If not, then the second expression alone matches both of them.

Side note: Matlab does not support hex/binary floating point literals, which is why the first expression above is limited to decimal digits.
Also, Matlab does not support capital I or J for imaginary numbers, nor underscore separators; however, the resulting token is invalid in Matlab, it's not a valid but different token, so highlighting it as well is not very harmful.
Yet another Matlab quirk: suffixes are only valid for hex/binary literals, not for decimal ones. So, striving for a perfect replication of Matlab's own highlighting would be quite complex.

**Comment by mpheath** [comment]
Could extend the not !(...) code and change:

https://github.com/ScintillaOrg/lexilla/blob/b2e3d5eb629e3937a02e4bc5d845acd9ab3b94a3/lexers/LexMatlab.cxx#L287-L292

to

```c++
			if (!isdigit(sc.ch) && sc.ch != '.'
			        && !(sc.ch == 'e' || sc.ch == 'E')
			        && !((sc.ch == '+' || sc.ch == '-') && (sc.chPrev == 'e' || sc.chPrev == 'E'))
			        && !(((sc.ch == 'x' || sc.ch == 'X') && sc.chPrev == '0') || (sc.ch >= 'a' && sc.ch <= 'f') || (sc.ch >= 'A' && sc.ch <= 'F'))
			        && !(sc.ch == 's' || sc.ch == 'S' || sc.ch == 'u' || sc.ch == 'U')
			        && !(sc.ch == 'i' || sc.ch == 'I' || sc.ch == 'j' || sc.ch == 'J')) {
				sc.SetState(SCE_MATLAB_DEFAULT);
				transpose = true;
			}
```

No change state is performed so it colours the numeric literal until it fails the match and then sets state to default and continues. So, the user should see the mistake if a typo is done. Example, last test line `c = 1e2j;` colours OK though `c = 1e2jZ;` with the appended `Z` will colour the `Z` as default style.

Checks on digits `0-9` or `.` is futile as setting to default will be set as numeric again at bottom of the loop and so checking digits after `s` or `u` and setting to default will just be set back to numeric.

<details>

<summary>Click to view image</summary>

![matlab](https://user-images.githubusercontent.com/58158242/206968250-65655152-0211-45a5-997d-a7c2134b074b.gif)

<hr>

</details>

If considered OK: [matlab_numeric_literals1.zip](https://github.com/ScintillaOrg/lexilla/files/10204815/matlab_numeric_literals1.zip)

Note: I have no experience with MatLab. Just recently browsed some docs.

**Comment by nyamatongwe** [comment]
The extended set of characters looked good so committed.

**Comment by zufuliu** [comment]
@nyamatongwe I think it's better to add milestone to issue or pull request that has been committed.
https://github.com/ScintillaOrg/lexilla/milestones

**Comment by nyamatongwe** [comment]
Milestones don't really match how things occur in Lexilla - there are no goals to have some set of bug fixes or features in a particular release. Its not like a commercial product where you have a set of resources that can be dispatched to progress the code towards multiple releases.

**Comment by zufuliu** [comment]
The goal is not to add bug or feature to a release (it's absolutely unknown what will be included), but save your time to write and reply "Included in 5.2.1 release.". There will be two useful milestones summary pages:
 
![Milestones](https://user-images.githubusercontent.com/2289926/211194842-be2e851d-e3ba-4cea-8e51-4e63b6b4e8ec.png)

![Milestones-issues](https://user-images.githubusercontent.com/2289926/211194913-e43f06ed-90c9-41b5-bcfe-83dccd353730.png)

It's easy to see what was resolved in a release (though read LexillaHistory.html is still required for downstream project developers).

**Comment by zufuliu** [comment]
Edited previous comment: added third screenshot and changed `it's absolutely unknown what will be included` to `it's absolutely unknown what will be included before the patch is applied` to make it clean that milestone is only added to issue or pull request that have patch applied or merged, not used for assigning task to the community.

You can add me as member or collaborator, so I can take time to add milestone to closed issues, it's still useful as some project (has habit to) not use latest release (e.g. Notepad++), or for browsing issues on GitHub.

The discussion for adding milestone can be spit out as an infrastructure issue. sorry for my bad to not start new issue at the begging.

**Comment by nyamatongwe** [comment]
Fixed in 5.2.2.
